### PR TITLE
`numpy.clip` accept `numpy.array` for `a_min`, `a_max`

### PR DESCRIPTION
**Pull request by czgdp1807**
Closes https://github.com/numba/numba/issues/7521

I have used `np.broadcast_arrays` to support arrays in `a_min` and `a_max` arguments of `np.clip`. Note that this doesn't deal with supporting scalars in `a` argument. However, I can do that if required. 

cc: @guilhermeleobas @gmarkall @Twomem

### CODE REVIEW

**Comment by Twomem** [comment]
> Closes #7521
> 
> I have used `np.broadcast_arrays` to support arrays in `a_min` and `a_max` arguments of `np.clip`. Note that this doesn't deal with supporting scalars in `a` argument. However, I can do that if required.
> 
> cc: @guilhermeleobas @gmarkall @Twomem

Hi @czgdp1807 , just checking, is the plan still to also add similar support for numpy.minimum? Not sure if this is in this pull request or it would be better to do in a separate request. Thanks!!

**Comment by czgdp1807** [comment]
> is the plan still to also add similar support for numpy.minimum?

I can do that, just waiting for reviews on this PR.

File: numba/tests/test_array_methods.py
Comment (guilhermeleobas) [inline numba/tests/test_array_methods.py:1455]: Move this to its own function.

File: numba/np/arrayobj.py
Comment (guilhermeleobas) [inline numba/np/arrayobj.py:1988]: Rename this function to something that reflects that `a_min` is None here.
```suggestion
def _np_clip_impl_amin_none(a, a_min, a_max, out):
```

File: numba/np/arrayobj.py
Comment (guilhermeleobas) [inline numba/np/arrayobj.py:1998]: This can be rewritten to 
```suggestion
        ret[index] = min(val_a_max, val_a)
```

File: numba/np/arrayobj.py
Comment (guilhermeleobas) [inline numba/np/arrayobj.py:2014]:
```suggestion
        ret[index] = max(val_a_min, val_a)
```

File: numba/np/arrayobj.py
Comment (guilhermeleobas) [inline numba/np/arrayobj.py:2004]: Same here. Replace the name to something that says `a_max` is None:
```suggestion
def _np_clip_impl_amax_none(a, a_min, a_max, out):
```

File: numba/np/arrayobj.py
Comment (guilhermeleobas) [inline numba/np/arrayobj.py:2060]: Since both `a_min` and `a_max` are scalars, I think it will be faster to compute the result without broadcasting them to arrays.

File: numba/np/arrayobj.py
Comment (guilhermeleobas) [inline numba/np/arrayobj.py:2073]:
```suggestion
                return _np_clip_impl_amax_none(a, a_min_full, a_max, out)
```

File: numba/np/arrayobj.py
Comment (guilhermeleobas) [inline numba/np/arrayobj.py:2096]:
```suggestion
                return _np_clip_impl_amin_none(a, a_min, a_max_full, out)
```

File: numba/np/arrayobj.py
Comment (guilhermeleobas) [inline numba/np/arrayobj.py:1998]: One can also simplify both `np_clip_impl_none1` and `np_clip_impl_none2` with

```python
@register_jitable
def _np_clip_impl_none(a, b, use_min, out):
    for index in np.ndindex(a.shape):
        val_a = a[index]
        val_b = b[index]
        if use_min:
            out[index] = min(val_a, val_b)
        else:
            out[index] = max(val_a, val_b)
    return out


@register_jitable
def _np_clip_impl_amin_none(a, a_min, a_max, out):
    # a_min is None but a_max is a numpy array
    ret = np.empty_like(a) if out is None else out
    a_b, a_max_b = np.broadcast_arrays(a, a_max)
    return _np_clip_impl_none(a_b, a_max_b, True, ret)


@register_jitable
def _np_clip_impl_amax_none(a, a_min, a_max, out):
    # a_min is a numpy array but a_max is None
    ret = np.empty_like(a) if out is None else out
    a_b, a_min_b = np.broadcast_arrays(a, a_min)
    return _np_clip_impl_none(a_b, a_min_b, False, ret)
```

**Review by guilhermeleobas — CHANGES_REQUESTED** [review]
Hi @czgdp1807, thanks for the patch. I've done an initial review.

File: numba/np/arrayobj.py
Comment (guilhermeleobas) [inline numba/np/arrayobj.py:2113]: I guess you can just copy the body of `_np_clip_impl_amin_none` here and remove the jitted function.

File: numba/np/arrayobj.py
Comment (guilhermeleobas) [inline numba/np/arrayobj.py:2118]: Same here. Since `_np_clip_impl_amax_none` is only used here. We can avoid the cost of jitting another function by copying the body of the function into here.

**Review by guilhermeleobas — CHANGES_REQUESTED** [review]
Thanks, @czgdp1807. Only a couple more requests.

**Comment by czgdp1807** [comment]
@guilhermeleobas I have addressed your latest reviews to inline jitted functions which are called only once. Please let me know if there is anything else to be done here.

**Review by guilhermeleobas — APPROVED** [review]
LGTM. Thanks, @czgdp1807

**Comment by czgdp1807** [comment]
Hi. @sklam @stuartarchibald @esc Is it ready for merging? Please let me know if anything is left to be done here. Thanks.

**Comment by czgdp1807** [comment]
Hi. I am starting my work on supporting arrays in `np.minimum` as this PR has received one approval and no objections. So, I think there is support for this work. Please let me know if the case is opposite. Thanks.

**Review by sklam — APPROVED** [review]
LGTM

**Comment by Twomem** [comment]
> Hi. I am starting my work on supporting arrays in `np.minimum` as this PR has received one approval and no objections. So, I think there is support for this work. Please let me know if the case is opposite. Thanks.
> 
> Edit - Seems like `np.minimum` already supports arrays as inputs (may be because its a `ufunc`). @Twomem Please let me know if you have any function in mind which should support arrays as inputs but currently doesn't. Thanks.

Thanks, @czgdp1807 -- I think if all the ufuncs are supported that would be the main thing as from what I can see, the biggest performance gains can be had using vectorise / parallel / fast math and GPU acceleration all benefit most from the ufuncs. What I don't get is whether np.clip is a ufunc -- in my mind it should be but the docs don't seem to list it

**Pull request merged**
